### PR TITLE
[Core] Making container final to avoid derivation

### PR DESCRIPTION
**Pull request by loumalouomega**
**Description**
Making container final to avoid derivation

**Changelog**
- Making container final to avoid derivation

### CODE REVIEW

**Comment by loumalouomega** [comment]
@KratosMultiphysics/technical-committee

**Comment by maceligueta** [comment]
Has this PR relationship with what I reported about the method 'Position' not being inlined by the compiler?
https://github.com/KratosMultiphysics/Kratos/blob/83f8ed11cd0ac57066daa6f5c8b27f0f66ce95bd/kratos/containers/variables_list_data_value_container.h#L933

**Comment by loumalouomega** [comment]
> Has this PR relationship with what I reported about the method 'Position' not being inlined by the compiler?
> 
> https://github.com/KratosMultiphysics/Kratos/blob/83f8ed11cd0ac57066daa6f5c8b27f0f66ce95bd/kratos/containers/variables_list_data_value_container.h#L933
> 
> The compiler did inline it when I switched to __forceinline, but maybe removing some virtuals does the work.

Yes, is the best way to avoid this (the compiler will not allow it)

**Comment by philbucher** [comment]
@maceligueta how do you check if it actually gets inlined?

**Comment by maceligueta** [comment]
I am trusting Vtune. There's an option to show/hide inline methods. It worked with the [] operator of Geometry.h. When I removed the virtuals, the operator never showed up again in the profiler results, not even as an inline method.

**Comment by loumalouomega** [comment]
The point of this PR is to make impossible such error

File: kratos/containers/all_variables_data_value_container.h [hunk 1]
```diff
@@ -55,7 +55,7 @@ namespace Kratos
 /// Short class definition.
 /** Detail class definition.
 */
-class KRATOS_EXPORT_DLL AllVariablesDataValueContainer
```
Comment (RiccardoRossi) [inline kratos/containers/all_variables_data_value_container.h:58]: @pooyan-dadvand are u ok with the change to this file?

File: kratos/containers/pointer_hash_map_set.h [hunk 2]
```diff
@@ -64,7 +64,7 @@ template<class TDataType,
          class THashType = std::hash<TDataType>,
          class TGetKeyType = SetIdentityFunction<TDataType>,
          class TPointerType = Kratos::shared_ptr<TDataType> >
-class PointerHashMapSet
```
Comment (RiccardoRossi) [inline kratos/containers/pointer_hash_map_set.h:67]: @pooyan-dadvand we need your comment on the changes in the base containers (which make sense to us).

specifically i am ok with the changes to csr related stuff

File: kratos/containers/distributed_vector_exporter.h [hunk 3]
```diff
@@ -53,7 +50,7 @@ namespace Kratos
 
 /// Provides a DistributedVectorExporter which implements FEM assemble capabilities
 template<class TIndexType=std::size_t>
-class DistributedVectorExporter
+class DistributedVectorExporter final
```
Comment (RiccardoRossi) [inline kratos/containers/distributed_vector_exporter.h:53]: i see no strong reason for making this class underivable, nevertheless it is ok to me that if someone wants to derive will eventually have to remove the final so this gets communicated

Comment (RiccardoRossi) [reply kratos/containers/distributed_vector_exporter.h:53]: (same comment for the importer)

**Comment by roigcarlo** [comment]
Why is `tree.h` final and for example, `octree.h` or `kd_tree.h` are not?

**Comment by loumalouomega** [comment]
Ping

**Comment by loumalouomega** [comment]
> Why is `tree.h` final and for example, `octree.h` or `kd_tree.h` are not?

I did not implement that, i don't know

**Comment by pooyan-dadvand** [comment]
I am not against it but I am not sure what is the benefit. Would you please add the benefits to the PR description for future references? (In case of a request for deriving one of those classes)

@maceligueta, About the virtual, making it final won't help inlining unless the class is not derived (which making final does not have sense neither)

**Comment by RiccardoRossi** [comment]
we had a short discussion in @KratosMultiphysics/technical-committee and we are in principle in agreement about this change.

The rationale about our support is that we can eventually easily "revert" this change and making it deliverable if we want to do it.
on the other hand adding the "final" may have some performance advantage as well as a small reduction (the pointer to the vtable) in the memory occupation of the object.

we will try to review it shortly. Adding @pooyan-dadvand and @RiccardoRossi as assignees

**Review by rubenzorrilla — APPROVED** [review]
I've carefully reviewed all the files and the changes are correct. I think that all these containers must  be `final`, specially considering that it would help the compiler in the devirtualization. If we eventually need to derive one the `final` specifier can be easily removed.

Approving on behalf of the @KratosMultiphysics/technical-committee . @loumalouomega feel free to merge. Thanks!

**Pull request merged**